src/components/main-panel/main-panel.tsx

```diff
@@ -108,38 +116,107 @@ export const MainPanel = () => {
         }
     };
 
-    const handleDragEnd = (event: DragEndEvent) => {
-        const { active, over } = event;
+    const handleOnDragStart = (event: DragStartEvent) => {
+        dragStartY.current = event.active.rect.current.initial?.top;
+        elementRefs.current.forEach((elementRef, index) => {
+            elementStartingRefs.current[index] =
+                elementRef.element.getBoundingClientRect();
+        });
+    };
 
-        if (active.id !== over.id) {
-            const oldIndex = orderedIdentifiers.findIndex(
-                (item) => item.internalId === active.id
-            );
-            const newIndex = orderedIdentifiers.findIndex(
-                (item) => item.internalId === over.id
+    const handleOnDragMove = debounce((event: DragMoveEvent) => {
+        const { active } = event;
+
+        const wrapperBounds = wrapperRef.current?.getBoundingClientRect();
+
+        const wrapperLeftThirds = wrapperBounds.left + wrapperBounds.width / 3;
+        const wrapperRightThirds =
+            wrapperBounds.left +
+            (wrapperBounds.width * 2) / 3 -
+            wrapperBounds.width / 9;
+
+        const currentActiveIndex = orderedIdentifiers.findIndex(
+            (item) => item.internalId === active.id
+        );
+
+        const mouseX =
+            active.rect.current.translated?.left +
+            event.activatorEvent["layerX"];
+
+        let elementIndex = currentActiveIndex;
+        let size: IElementIdentifier["size"] = "full";
+        if (mouseX <= wrapperLeftThirds) {
+            size = "left";
+        } else if (mouseX >= wrapperRightThirds) {
+            size = "right";
+        }
+
+        if (active.rect.current.translated.top > dragStartY.current) {
+            // move down
+            elementIndex = elementStartingRefs.current.findLastIndex(
+                (elementRect) =>
+                    active.rect.current.translated.bottom + 16 >
+                    elementRect.y + elementRect.height / 2
```

Comment:
notice the build is failing with error for `findLastIndex`, it appears the pipelines do not support this yet so we shall use `findIndex()`

```suggestion
            elementIndex = elementStartingRefs.current.toReversed().findIndex(
                (elementRect) =>
                    active.rect.current.translated.bottom + 16 >
                    elementRect.y + elementRect.height / 2
```